File: Grafo.java
public class Grafo {

    public static final int INF = 999999;
    public static final int MAX = 100;

    private String[] ciudades;
    private int[][]  matriz;
    private int      numVertices;

    public Grafo() {
        ciudades    = new String[MAX];
        matriz      = new int[MAX][MAX];
        numVertices = 0;
    }

    public boolean agregarNodo(String ciudad) {
        if (buscarIndice(ciudad) >= 0) return false;
        for (int i = 0; i <= numVertices; i++) {
            matriz[numVertices][i] = INF;
            matriz[i][numVertices] = INF;
        }
        matriz[numVertices][numVertices] = 0;
        ciudades[numVertices] = ciudad;
        numVertices++;
        return true;
    }

    public boolean agregarArco(String c1, String c2, int km) {
        agregarNodo(c1);
        agregarNodo(c2);
        matriz[buscarIndice(c1)][buscarIndice(c2)] = km;
        return true;
    }

    public boolean eliminarArco(String c1, String c2) {
        int i = buscarIndice(c1);
        int j = buscarIndice(c2);
        if (i < 0 || j < 0 || matriz[i][j] == INF) return false;
        matriz[i][j] = INF;
        return true;
    }

    public boolean existeArco(String c1, String c2) {
        int i = buscarIndice(c1);
        int j = buscarIndice(c2);
        return i >= 0 && j >= 0 && matriz[i][j] != INF;
    }

    public int getNumVertices()       { return numVertices; }
    public String getCiudad(int i)    { return ciudades[i]; }
    public boolean existeCiudad(String c) { return buscarIndice(c) >= 0; }

    public int getIndice(String ciudad) { return buscarIndice(ciudad); }

    public int[][] getMatriz() {
        int[][] copia = new int[numVertices][numVertices];
        for (int i = 0; i < numVertices; i++)
            for (int j = 0; j < numVertices; j++)
                copia[i][j] = matriz[i][j];
        return copia;
    }

    public String[] listarCiudades() {
        String[] lista = new String[numVertices];
        for (int i = 0; i < numVertices; i++) lista[i] = ciudades[i];
        return lista;
    }

    public void imprimirMatriz() {
        System.out.println("\n===== MATRIZ DE ADYACENCIA =====");
        System.out.printf("%-20s", "");
        for (int j = 0; j < numVertices; j++)
            System.out.printf("%-14s", cortar(ciudades[j], 12));
        System.out.println();
        for (int i = 0; i < numVertices; i++) {
            System.out.printf("%-20s", cortar(ciudades[i], 18));
            for (int j = 0; j < numVertices; j++)
                System.out.printf("%-14s", matriz[i][j] == INF ? "INF" : matriz[i][j]);
            System.out.println();
        }
        System.out.println("================================\n");
    }

    private int buscarIndice(String ciudad) {
        for (int i = 0; i < numVertices; i++)
            if (ciudades[i].equals(ciudad)) return i;
        return -1;
    }

    private String cortar(String s, int max) {
        return s.length() > max ? s.substring(0, max - 2) + ".." : s;
    }
}

File: grafo.py
# grafo.py
# Wrapper sobre NetworkX para representar el grafo dirigido de Guatemala

import networkx as nx

class Grafo:
    def __init__(self):
        self.G = nx.DiGraph()

    def agregar_nodo(self, ciudad):
        if ciudad not in self.G:
            self.G.add_node(ciudad)
            return True
        return False

    def agregar_arco(self, c1, c2, km):
        self.G.add_edge(c1, c2, weight=km)

    def eliminar_arco(self, c1, c2):
        if self.G.has_edge(c1, c2):
            self.G.remove_edge(c1, c2)
            return True
        return False

    def existe_arco(self, c1, c2):
        return self.G.has_edge(c1, c2)

    def existe_ciudad(self, ciudad):
        return ciudad in self.G

    def listar_ciudades(self):
        return list(self.G.nodes())

    def num_ciudades(self):
        return self.G.number_of_nodes()

    def imprimir_matriz(self):
        nodos = list(self.G.nodes())
        ancho = max((len(n) for n in nodos), default=5) + 2
        col   = 10
        print("\n===== MATRIZ DE ADYACENCIA =====")
        print(f"{'':>{ancho}}", end="")
        for n in nodos:
            print(f"{n:>{col}}", end="")
        print()
        for origen in nodos:
            print(f"{origen:>{ancho}}", end="")
            for destino in nodos:
                if origen == destino:
                    print(f"{'0':>{col}}", end="")
                elif self.G.has_edge(origen, destino):
                    print(f"{self.G[origen][destino]['weight']:>{col}}", end="")
                else:
                    print(f"{'INF':>{col}}", end="")
            print()
        print("================================\n")
